### PR TITLE
 [Hardware][Ascend] MiniMax H3: route packed-sequence attention through npu_fused_infer_attention_score varlen 

### PR DESCRIPTION
**Pull request by Qeeweew**
## Purpose

Route NPU attention through `torch.ops.npu.npu_fused_infer_attention_score` (FIA)
in TND varlen mode when `VLLM_OMNI_NPU_FIA_VARLEN=1`, as an opt-in alternative
to the default `fused_attn_score` + full-QK-mask path in
`FlashAttentionImpl.forward_fa_npu`.

The default path materializes a `[B,1,Sq,Skv]` attention mask and operator
workspace that is O(S^2) in HBM and dominates peak memory for long packed
sequences. **MiniMax H3** packs text/audio/video latents into one sequence (S can
exceed 100k rows); 15 s 1344x768 requests failed with an NPU OOM at a 45 GiB
attention allocation before the first denoise step.

The FIA varlen path mirrors the CUDA `flash_attn_varlen_func` contract used by
`_forward_varlen_packed`: packed (T, N, D) rows with vLLM cu_seqlens document
boundaries mapped directly onto the op's `actual_seq_lengths` (prefix-accumulated
with leading 0, segment length = adjacent difference). torch_npu accepts device
int32 tensors there, so no host-side copy or sync is introduced per call. It
covers any batch size / document count and falls back to the default path on
contract mismatch (input rank / missing boundaries).

Off by default: it changes the attention operator and may shift numerics at
~1e-3.

## Test Plan

**vLLM Version:** 0.26.0

**vLLM-Omni Commit:** b55ab5f7 (branch `feat/npu-fia-varlen-attention`)

Environment: 8 x 910B3 (CANN 9.0.1, torch_npu 2.10.0.post4),
MiniMax-H3 FL2VA partition.

- Single-op correctness vs PyTorch SDPA (bf16): two-segment, three-segment,
  single-segment, and multi-batch cross-batch cu_seqlens scenarios, plus
  56-head no-pad (token_refiner) — all max_err < 5e-3.
- Peak-memory scaling at H3 shape (N=7, D=128): S=109,912 attention peak
  increment ~2.9 GiB vs 45 GiB (S^2 x 4B) on the default path.
- End-to-end: T2VA 15 s 1344x768 (previously OOM at first denoise step) now
  completes; output verified 1344x768@24fps / 362 frames / 32 kHz stereo.

## Test Result

- Single-op script: `scripts/test_fia_sem.py`, `scripts/test_fia_generic.py`
  (all PASS)
- E2E: `videos/t2va_15s_768p_fia.mp4` generated without OOM;

### CODE REVIEW

File: vllm_omni/diffusion/attention/backends/flash_attn.py [hunk 1]
```diff
@@ -417,3 +434,38 @@ def forward_fa_npu(
...
+
+        q = query.reshape(-1, N, D)
+        k = key.reshape(key.shape[0] * key.shape[1], N, D)
+        v = value.reshape(value.shape[0] * value.shape[1], N, D)
+        out = torch.ops.npu.npu_fused_infer_attention_score(
+            q,
+            k,
+            v,
+            num_heads=N,
+            num_key_value_heads=N,
```
Comment (david6666666) [inline vllm_omni/diffusion/attention/backends/flash_attn.py]: This hardcodes MHA (kv heads == query heads), and the k/v reshape above requires key.shape[2] == N, so a GQA model would raise instead of falling back -- contradicting the PR's 'falls back on contract mismatch' claim. MiniMax-H3 is MHA so it works, but this is a global opt-in for all NPU models. Please derive kv heads from key.shape[2] (and pass that as num_key_value_heads) so GQA either works or is treated as a contract mismatch.

File: vllm_omni/diffusion/attention/backends/flash_attn.py [hunk 2]
```diff
@@ -417,3 +434,38 @@ def forward_fa_npu(
...
+            scale=self.softmax_scale,
+            input_layout="TND",
+            actual_seq_lengths=cu_seqlens_q,
+            actual_seq_lengths_kv=cu_seqlens_kv,
+            sparse_mode=0,
```
Comment (david6666666) [inline vllm_omni/diffusion/attention/backends/flash_attn.py:479]: sparse_mode=0 with no atten_mask is dense attention within each segment, and this path never consults self.causal or attn_metadata.attn_mask. For MiniMax-H3 that matches the reference (causal=False), but note the model builds a mask when used < packed_total (the alignment-padding second doc); here that padding doc attends to itself instead of being masked out -- harmless only because those rows are discarded downstream. Since this env var is a global opt-in, please gate the fast path on not self.causal and a trivial/absent mask so a causal or masked NPU model does not silently lose its mask.

**Comment by david6666666** [comment]
please add perf and accuracy comparasion

File: vllm_omni/diffusion/attention/backends/flash_attn.py (hunk 1 above)
Comment (Qeeweew) [reply vllm_omni/diffusion/attention/backends/flash_attn.py]: 改了

**Comment by Qeeweew** [comment]
```python
"""长序列单算子汇总：FIA varlen 精度 + TFLOPS，原路径（fused_attn_score+mask）对照。"""
import time, torch, torch_npu
torch_npu.npu.set_device(0)
torch.manual_seed(0)
FIA = torch.ops.npu.npu_fused_infer_attention_score
from mindiesd import attention_forward

def tflops(T, N, D, dt_s):
    return 4.0 * N * T * T * D / dt_s / 1e12

def bench_fia(T, N, D, cu, iters=30, warm=5):
    q = torch.randn(T, N, D, dtype=torch.bfloat16, device="npu")
    k = torch.randn(T, N, D, dtype=torch.bfloat16, device="npu")
    v = torch.randn(T, N, D, dtype=torch.bfloat16, device="npu")
    cu_t = torch.tensor(cu, dtype=torch.int32, device="npu")
    def run():
        return FIA(q, k, v, num_heads=N, num_key_value_heads=N, scale=D**-0.5,
                   input_layout="TND", actual_seq_lengths=cu_t,
                   actual_seq_lengths_kv=cu_t, sparse_mode=0)[0]
    for _ in range(warm): run()
    torch_npu.npu.synchronize()
    t0 = time.perf_counter()
    for _ in range(iters): run()
    torch_npu.npu.synchronize()
    return (time.perf_counter() - t0) / iters

def bench_fused(T, N, D, iters=10, warm=3):
    """原路径：fused_attn_score + full-QK mask（flash_attn.py 默认路径，BNSD 4D）。"""
    q = torch.randn(1, T, N, D, dtype=torch.bfloat16, device="npu")
    k = torch.randn(1, T, N, D, dtype=torch.bfloat16, device="npu")
    v = torch.randn(1, T, N, D, dtype=torch.bfloat16, device="npu")
    mask = torch.ones(1, 1, T, T, dtype=torch.bool, device="npu")
    def run():
        return attention_forward(q, k, v, attn_mask=mask, opt_mode="manual",
                                 op_type="fused_attn_score", layout="BNSD")
    for _ in range(warm): run()
    torch_npu.npu.synchronize()
    t0 = time.perf_counter()
    for _ in range(iters): run()
    torch_npu.npu.synchronize()
    return (time.perf_counter() - t0) / iters

N, D = 7, 128
print("=" * 62)
print(f"长序列: T=109,912 N={N} D={D}（H3 主序列 15s 768P，两段 text+media）")
print("=" * 62)
# 1) 显存/可行性：原路径在长序列上
try:
    dt = bench_fused(109_912, N, D, iters=2, warm=1)
    print(f"[原路径 fused_attn_score+mask] 跑通 dt={dt*1e3:.1f}ms")
except Exception as e:
    print(f"[原路径 fused_attn_score+mask] 长序列 OOM/失败: {str(e).splitlines()[0][:90]}")
# 2) FIA 长序列性能
dt = bench_fia(109_912, N, D, [0, 100, 109_912])
print(f"[FIA varlen] 长序列 dt={dt*1e3:.1f}ms  TFLOPS={tflops(109_912, N, D, dt):.1f}")

print()
print("=" * 62)
print(f"同规模对比（原路径可跑的最大规模附近）: T=16,384")
print("=" * 62)
dt_fia = bench_fia(16_384, N, D, [0, 100, 16_384])
print(f"[FIA varlen]      dt={dt_fia*1e3:.1f}ms   TFLOPS={tflops(16_384, N, D, dt_fia):.1f}")
dt_fu = bench_fused(16_384, N, D)
print(f"[原路径+mask]     dt={dt_fu*1e3:.1f}ms   TFLOPS={tflops(16_384, N, D, dt_fu):.1f}   (峰值~{4*16_384**2*2/2**30:.1f}GiB workspace)")

print()
print("=" * 62)
print("精度（vs PyTorch SDPA 逐段，两段式长序列 T=16,384）")
print("=" * 62)
T = 16_384
q = torch.randn(T, N, D, dtype=torch.bfloat16, device="npu")
k = torch.randn(T, N, D, dtype=torch.bfloat16, device="npu")
v = torch.randn(T, N, D, dtype=torch.bfloat16, device="npu")
cu = [0, 100, T]
out = FIA(q, k, v, num_heads=N, num_key_value_heads=N, scale=D**-0.5,
          input_layout="TND", actual_seq_lengths=cu,
          actual_seq_lengths_kv=cu, sparse_mode=0)[0]
for i, (s, e) in enumerate([(0, 100), (100, T)]):
    ref = torch.nn.functional.scaled_dot_product_attention(
        q[s:e].float().transpose(0,1).unsqueeze(0),
        k[s:e].float().transpose(0,1).unsqueeze(0),
        v[s:e].float().transpose(0,1).unsqueeze(0),
        scale=D**-0.5, is_causal=False).squeeze(0).transpose(0,1)
    err = (out[s:e].float() - ref).abs().max().item()
    print(f"  段{i} [{s}:{e}] len={e-s}: max err={err:.2e} {'PASS' if err<5e-3 else 'FAIL'}")

```

910B-3结果：

```
==============================================================
长序列: T=109,912 N=7 D=128（H3 主序列 15s 768P，两段 text+media）
==============================================================
[原路径 fused_attn_score+mask] 跑通 dt=458.3ms
[FIA varlen] 长序列 dt=445.2ms  TFLOPS=97.3

==============================================================
同规模对比（原路径可跑的最大规模附近）: T=16,384
==============================================================
[FIA varlen]      dt=7.8ms   TFLOPS=123.5
[原路径+mask]     dt=9.5ms   TFLOPS=101.3   (峰值~2.0GiB workspace)

==============================================================
精度（vs PyTorch SDPA 逐段，两段式长序列 T=16,384）
==============================================================
  段0 [0:100] len=100: max err=3.75e-03 PASS
  段1 [100:16384] len=16284: max err=2.51e-04 PASS

```

这里只能单算子简单测试一下

**Comment by Qeeweew** [comment]
其实MiniMax H3这边应该可以不用npu_fused_infer_attention_score，理论上裁剪一下就行了，这样就不用物化mask，我先关闭了
